### PR TITLE
Fix NACA

### PR DESCRIPTION
**Pull request by TheBB**
Something must have changed.

### CODE REVIEW

**Comment by TheBB** [comment]
@VikingScientist Did I get the periodicity right?

File: splipy/utils/NACA.py [hunk 1]
```diff
@@ -82,11 +82,19 @@ def thickness(x):
         a1 = -0.126
         a2 = -0.3516
         a3 = 0.2843
-        a4 = -0.1015
+        a4 = -0.1036 if closed else -0.1015
```
Comment (VikingScientist) [inline splipy/utils/NACA.py:85]: Why is this needed?

**Comment by VikingScientist** [comment]
Yes, you did get the periodicity right, though I would like it all wrapped into a nice `Curve.makePeriodic` function with the only change from the proposed logic: it should change both the end and start controlpoint to be the average of the two.

Also: `surface_factory.thicken()` was requested rewritten by PR #15 , and I see that the NACA script relies on this. No immediate action need to be taken, but it is nice to be aware of.

**Comment by VikingScientist** [comment]
Also: since you're looking into this now. Fix the typos in the documentation (i.e. Line 60: `@param P`)

File: splipy/utils/NACA.py (hunk 1 above)
Comment (TheBB) [reply splipy/utils/NACA.py:85]: http://airfoiltools.com/airfoil/naca4digit

I got it from here.

Comment (VikingScientist) [reply splipy/utils/NACA.py:85]: Would you look at that. The default NACA profiles aren't even defined as closed loops :fearful:

**Comment by TheBB** [comment]
Okay, new commits with tests. Please double check though.

File: splipy/BSplineBasis.py [hunk 2]
```diff
@@ -304,6 +319,8 @@ def raise_order(self, amount):
             raise TypeError('amount needs to be a non-negative integer')
         if amount < 0:
             raise ValueError('amount needs to be a non-negative integer')
+        if amount == 0:
+            return self.clone()
```
Comment (TheBB) [inline splipy/BSplineBasis.py:323]: If you're wondering about this, it's because I wasn't able to `raise_order(0)` on a periodic basis, which I found it easier (for now) to single out the trivial case rather than fix.

File: splipy/SplineObject.py [hunk 3]
```diff
@@ -1147,6 +1147,49 @@ def split(self, knots, direction=0):
...
+        if not 0 <= continuity <= basis.order - 2:
+            raise ValueError('Illegal continuity for basis of order {}: {}'.format(
+                continuity, order
+            ))
+        if basis.periodic >= 0:
+            raise ValueError('Basis is already periodic')
+
+        basis = basis.make_periodic(continuity)
+
+        # Merge control points
```
Comment (VikingScientist) [inline splipy/SplineObject.py:1175]: Think this will fail for rational controlpoints

File: splipy/utils/NACA.py [hunk 4]
```diff
@@ -53,18 +53,18 @@ def camber(M, P, order=5):
...
-    @type  n: Int
-    @param order: Order of resulting spline curve
-    @type  order: Int
-    @return : Approximation of NACA wing profile
-    @rtype  : Curve
+    :param M: Max camber height (y) given as percentage 0% to 9% of length
+    :type  M: Int 0<M<10
+    :param P: Max camber position (x) given as percentage 0% to 90% of length
+    :type  P: Int 0<P<10
+    :param T: Thickness given in percentage 1
```
Comment (VikingScientist) [inline splipy/utils/NACA.py]: The function takes in an `X`

File: splipy/IO/svg.py [hunk 5]
```diff
@@ -312,13 +312,9 @@ def curve_from_path(self, path):
                     controlpoints.append(cp)
                 curve_piece = Curve(BSplineBasis(2, knot), controlpoints)
             elif piece[0] == 'z' or piece[0] == 'Z':
-                # manually set to periodic curve, since we don't have a Curve.make_periodic()
```
Comment (VikingScientist) [inline splipy/IO/svg.py:315]: :smile:  :tada:

File: splipy/BSplineBasis.py (hunk 2 above)
Comment (VikingScientist) [reply splipy/BSplineBasis.py:323]: I don't mind this.

File: splipy/SplineObject.py (hunk 3 above)
Comment (TheBB) [reply splipy/SplineObject.py:1175]: Is there something sensible to do for rationals?

Comment (VikingScientist) [reply splipy/SplineObject.py:1175]: Initial thoughts is to just treat the weights as just another controlpoint, i.e. use (self.pardim+self.rational) instead of self.pardim. Can't tell the immediate consequences. Might work. If not, then at least throw warning that this is not supported for rational input.

File: splipy/SplineObject.py [hunk 6]
```diff
@@ -1147,6 +1147,49 @@ def split(self, knots, direction=0):
...
+
+        :param continuity: The continuity along the boundary (default max).
+        :param direction: The direction to ensure continuity in.
+        """
+
+        direction = check_direction(direction, self.pardim)
+        basis = self.bases[direction]
+        if continuity is None:
+            continuity = basis.order - 2
+        if not 0 <= continuity <= basis.order - 2:
```
Comment (VikingScientist) [inline splipy/SplineObject.py]: For input continuity=-1, then the right function is `SplineObject.split( )`. Might consider calling it directly from here or provide a gentle hint (error/warning message) as to where the user can find it.

File: splipy/SplineObject.py (hunk 3 above)
Comment (TheBB) [reply splipy/SplineObject.py:1175]: That should be what it's doing already, `self.pardim + 1` is the number of dimensions in the CP array, not the number of dofs per CP. The `index_beg` and `index_end` arrays have `:` entries for each dimension except the one being operated on, including the last dimension.

**Comment by TheBB** [comment]
Thanks, updated. :+1:

**Comment by VikingScientist** [comment]
```
from splipy import *                                                           
from splipy.IO import *                                                           
import splipy.curve_factory as cf                                                 
                                                                                  
crv = cf.circle()      # rational spline                                                           
crv = crv.split(0)    # make non-periodic                         
crv = crv.make_periodic()  # make C1-periodic             
                                                                                  
with G2("out.g2") as f:                                                           
    f.write(crv)                                                                  
```
not-circle

File: splipy/SplineObject.py (hunk 3 above)
Comment (VikingScientist) [reply splipy/SplineObject.py:1175]: Yes of course. You are right: it won't crash. Tried it on the default circle and the initial results are not exactly promising...

Comment (TheBB) [reply splipy/SplineObject.py:1175]: I had to update, turns out I was modifying the original CP array instead of a copy.

You're the expert on NURBS so I'll defer to you, but I'll just note that `cf.circle()` creates a C0-continuous object, and after `.split(0).make_periodic(0)` I seem to be getting back the identical one.

Comment (TheBB) [reply splipy/SplineObject.py:1175]: Oh, but `cf.circle()` has a C1-continuous version, I see that now. And I'm not getting the correct one back. Hmm.

**Comment by VikingScientist** [comment]
OK, so you can't make it C<sup>1</sup> if it initially was no more than C<sup>0</sup>. If you try `crv.make_periodic(0)`, then it all works just as it should.

However, if the original actually was C<sup>1</sup> and you try and rip it apart and put it together again, then this does not work
``` python
crv = cf.circle(type='p4c1')                                                      
print(crv)                 # prints C1-periodic quartic circle parametrization                                        
crv = crv.split(0)         # make non-periodic                                    
crv = crv.make_periodic(1) # make periodic again                                  
print(crv)                 # ideally should be identical to the one above.                                       
```

**Output**
```
p=5, [-1.57079633 -1.57079633  0.          0.          0.          1.57079633
  1.57079633  1.57079633  3.14159265  3.14159265  3.14159265  4.71238898
  4.71238898  4.71238898  6.28318531  6.28318531  6.28318531  7.85398163
  7.85398163], C1-periodic
[[ 1.         -0.35355339  1.        ]
 [ 1.          0.35355339  1.        ]
 [ 0.77614237  0.77614237  0.94280904]
 [ 0.35355339  1.          1.        ]
 [-0.35355339  1.          1.        ]
 [-0.77614237  0.77614237  0.94280904]
 [-1.          0.35355339  1.        ]
 [-1.         -0.35355339  1.        ]
 [-0.77614237 -0.77614237  0.94280904]
 [-0.35355339 -1.          1.        ]
 [ 0.35355339 -1.          1.        ]
 [ 0.77614237 -0.77614237  0.94280904]]
p=5, [-1.57079633 -1.57079633  0.          0.          0.          1.57079633
  1.57079633  1.57079633  3.14159265  3.14159265  3.14159265  4.71238898
  4.71238898  4.71238898  6.28318531  6.28318531  6.28318531  7.85398163
  7.85398163], C1-periodic
[[ 1.         -0.1767767   1.        ]
 [ 1.          0.1767767   1.        ]
 [ 0.77614237  0.77614237  0.94280904]
 [ 0.35355339  1.          1.        ]
 [-0.35355339  1.          1.        ]
 [-0.77614237  0.77614237  0.94280904]
 [-1.          0.35355339  1.        ]
 [-1.         -0.35355339  1.        ]
 [-0.77614237 -0.77614237  0.94280904]
 [-0.35355339 -1.          1.        ]
 [ 0.35355339 -1.          1.        ]
 [ 0.77614237 -0.77614237  0.94280904]]
```

Do we even care to fix this? I do feel that this is a very niche case with C^1-periodic rational splines, and in general one looses information with `make_periodic` (this will remove one or more functions from our spline space), and it is thus to be expected that the geometry will have to change under this operation. In this particular case it is theoretical possible to get back to the original C^1-periodicity, but I'm not convinced its worth the hassle to make it work. If you have the motivation, then by all means, have fun with it, but for my part it is sufficient to throw error message if continuity>0 and it is rational

**Comment by TheBB** [comment]
I was playing around with it but it's not worth spending time on this at the moment I think. For the updated (corrected) version of this PR, I should note that there are weighted averages of CPs that **will** restore the orginal CP array in the C1-continuous circle case, notably

```python
new_cp[0] = 0 × old_cp[0] + 1 × old_cp[-2]
new_cp[1] = 1 × old_cp[1] + 0 × old_cp[-1]
```

I don't know if this can be generalized. At any rate, I have updated with an error message in this case and filed this in the back of my mind.

**Comment by TheBB** [comment]
Don't merge this yet. Give me a sec.

**Comment by VikingScientist** [comment]
NP. It all looks good now, so we'll merge when you've reorganized stuff to your liking :smile:

**Comment by TheBB** [comment]
There, made it work. Black magic. No shame at all. :beers:

**Comment by VikingScientist** [comment]
:+1:

Dark sorcery approved :new_moon: :smiling_imp: :sparkles:
